File: codes/naeun/876_middleofthelinkedlist.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {
        // only one -> return itself
        if(head.next == null){
            return head;
        }
        // fast: go with 2 steps
        ListNode fast = head.next;
        // slow: go with 1 step
        ListNode slow = head;

        // if fast reaches the end or over the end, return the next node of slow
        while(!endCondition(fast)){
            fast = fast.next.next;
            slow = slow.next;
        }

        if(fast == null){
            return slow;
        }
        return slow.next;
    }

    private boolean endCondition(ListNode fast){
        return fast == null || fast.next == null;
    }
}

File: codes/beomjin/876_middleofthelinkedlist.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
  def middleNode(head: ListNode): ListNode = {
    var len = 0
    var curr = head
    while (curr != null) {
      len += 1
      curr = curr.next
    }

    var mid = len / 2
    curr = head
    for (_ <- 0 until mid) {
      curr = curr.next
    }

    val dummy = new ListNode(0)
    var tail = dummy
    while (curr != null) {
      tail.next = new ListNode(curr.x)
      tail = tail.next
      curr = curr.next
    }

    dummy.next
  }
}

class ListNodeIterator(head: ListNode) extends Iterator[Int] {
  private var current = head

  override def hasNext: Boolean = current != null

  override def next(): Int = {
    if (!hasNext) throw new NoSuchElementException
    val value = current.x
    current = current.next
    value
  }
}